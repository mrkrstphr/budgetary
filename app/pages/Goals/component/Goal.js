import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

const GoalStyles = styled.div`
  .title {
    margin-bottom: 0;
  }

  .bar {
    border-radius: 3px;
    height: 30px;
    margin: 4px auto;

    &.red {
      border: 1px solid darkred;

      .progress {
        background-color: darkred;
        background: repeating-linear-gradient(
          45deg,
          darkred,
          darkred 7px,
          firebrick 7px,
          firebrick 14px
        );
      }
    }

    &.green {
      border: 1px solid limegreen;

      .progress {
        background-color: limegreen;
        background: repeating-linear-gradient(
          45deg,
          limegreen,
          limegreen 7px,
          lightgreen 7px,
          lightgreen 14px
        );
      }
    }

    &.yellow {
      border: 1px solid gold;

      .progress {
        background-color: gold;
        background: repeating-linear-gradient(
          45deg,
          gold,
          gold 7px,
          yellow 7px,
          yellow 14px
        );
      }
    }

    .progress {
      border-radius: 3px;
      height: 100%;
    }
  }

  .details {
    align-items: center;
    display: flex;

    .account {
      flex: 1;
    }
  }
`;

function calculateGoalPercentage(goal) {
  if (goal.account.type === 'bank') {
    return Math.min(goal.account.currentBalance / goal.goalBalance, 1) * 100;
  }

  if (goal.account.type === 'liabilities') {
    // FIXME this assumes the goal balance is 0
    return Math.abs(
      Math.min(goal.account.currentBalance / goal.startingBalance, 1) * 100,
    );
  }

  return 0;
}

function getBarColor(goal) {
  const percentage = calculateGoalPercentage(goal);

  if (goal.account.type === 'bank') {
    if (percentage > 75) {
      return 'green';
    }

    if (percentage > 50) {
      return 'yellow';
    }
  }

  if (goal.account.type === 'liabilities') {
    if (percentage < 10) {
      return 'green';
    }
    if (percentage < 50) {
      return 'yellow';
    }
  }

  return 'red';
}

export default function Goal({ goal }) {
  return (
    <GoalStyles>
      <h3 className="title">{goal.description}</h3>
      <div className={`bar ${getBarColor(goal)}`}>
        <div
          className="progress"
          style={{ width: `${calculateGoalPercentage(goal)}%` }}
        />
      </div>
      <div className="details">
        <div className="account">Account: {goal.account.name}</div>
        <div className="amount">
          Goal:{' '}
          {new Intl.NumberFormat('en-US', {
            maximumFractionDigits: 2,
            minimumFractionDigits: 2,
          }).format(goal.goalBalance)}
        </div>
      </div>
    </GoalStyles>
  );
}

Goal.propTypes = {
  goal: PropTypes.shape({
    description: PropTypes.string.isRequired,
    account: PropTypes.shape({
      currentBalance: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      type: PropTypes.oneOf(['bank', 'liabilities', 'credit']),
    }).isRequired,
    startingBalance: PropTypes.number.isRequired,
    goalBalance: PropTypes.number.isRequired,
  }).isRequired,
};
