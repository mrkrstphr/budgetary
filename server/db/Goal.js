const makeUuid = require('uuid/v4');
const Account = require('./Account');
const { desnakeify, pickFirst, snakeify } = require('../lib');

class Goal {
  constructor(conn) {
    this.conn = conn;
    this.account = new Account(conn);
  }

  async createGoal(accountId, goal) {
    const startingBalance = await this.account.calculateCurrentBalance(
      accountId,
    );

    return desnakeify(
      pickFirst(
        this.conn('goals').insert(
          snakeify({ id: makeUuid(), accountId, startingBalance, ...goal }),
          '*',
        ),
      ),
    );
  }

  fetchGoals() {
    return desnakeify(
      this.conn('goals')
        .select('*')
        .orderBy('due_date', 'ASC'),
    );
  }
}

module.exports = Goal;
