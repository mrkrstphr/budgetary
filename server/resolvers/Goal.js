module.exports = {
  account({ account, accountId }, params, context) {
    if (account) {
      return account;
    }

    return context.dataloaders.accountById.load(accountId);
  },
  created({ createdAt }) {
    return createdAt;
  },
};
