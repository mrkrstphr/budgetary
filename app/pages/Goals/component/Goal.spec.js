import { shallow } from 'enzyme';
import React from 'react';
import Goal from './Goal';

describe('components/Goal', () => {
  describe('for account type = bank', () => {
    it('should render a 0% goal on a bank account with 0 balance', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Foo',
            account: {
              currentBalance: 0,
              name: 'My Bank',
              type: 'bank',
            },
            startingBalance: 0,
            goalBalance: 400,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('0%');
    });

    it('should render a 100% goal on a bank account with a matching goal balance', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Have lots-o-money',
            account: {
              currentBalance: 300,
              name: 'My Other Bank',
              type: 'bank',
            },
            startingBalance: 0,
            goalBalance: 300,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('100%');
    });

    it('should render a 100% goal when the balance is over the goal', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Break the bank',
            account: {
              currentBalance: 650,
              name: 'My Other, Other Bank',
              type: 'bank',
            },
            startingBalance: 0,
            goalBalance: 500,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('100%');
    });

    it('should render a 0% goal when the balance falls below the starting balance', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Survive',
            account: {
              currentBalance: 1,
              name: 'Some Other Bank',
              type: 'bank',
            },
            startingBalance: 3500,
            goalBalance: 15000,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('0%');
    });
  });

  describe('for account type = liabilities', () => {
    it('should render the correct percentage', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Make progress on student loans',
            account: {
              currentBalance: 2100,
              name: 'Student Loan X',
              type: 'liabilities',
            },
            startingBalance: 4000,
            goalBalance: 2000,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('50%');
    });

    it('should render 0% when the current balance matches the goal balance', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Pay off student loans',
            account: {
              currentBalance: 0,
              name: 'Student Loan X',
              type: 'liabilities',
            },
            startingBalance: 2900,
            goalBalance: 0,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('0%');
    });

    it('should render 0% when the current balance matches the goal balance that is not 0', () => {
      const component = shallow(
        <Goal
          goal={{
            description: 'Pay off (most of) student loans',
            account: {
              currentBalance: 500,
              name: 'Student Loan Y',
              type: 'liabilities',
            },
            startingBalance: 11000,
            goalBalance: 500,
          }}
        />,
      );
      const styles = component.find('.progress').get(0).props.style;
      expect(styles.width).toEqual('0%');
    });
  });
});
