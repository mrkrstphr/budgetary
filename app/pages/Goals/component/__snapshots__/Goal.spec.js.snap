// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Goal for account type = liabilities should render 0% when the current balance matches the goal balance that is not 0 1`] = `
<div
  className="Goal__GoalStyles-sc-1afoxc-0 jJbUAz"
>
  <h3
    className="title"
  >
    Pay off (most of) student loans
  </h3>
  <div
    className="bar green"
  >
    <div
      className="progress"
      style={
        Object {
          "width": "5%",
        }
      }
    />
  </div>
  <div
    className="details"
  >
    <div
      className="account"
    >
      Account: 
      Student Loan Y
    </div>
    <div
      className="amount"
    >
      Goal:
       
      500.00
    </div>
  </div>
</div>
`;
